toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1388 @@
+};
+
+// Define these callback array types to make the code easier to read.
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherCallback>>
+  ChangeCallbackArray;
+typedef nsTArray<nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>>
+  ErrorCallbackArray;
+
+static mozilla::LazyLogModule gNativeWatcherPRLog("NativeFileWatcherService");
+#define FILEWATCHERLOG(...)                                                    \
+  MOZ_LOG(gNativeWatcherPRLog, mozilla::LogLevel::Debug, (__VA_ARGS__))
+
+struct WatchedResourceDescriptor
+{
+  // The path on the file system of the watched resource.
+  nsString mPath;
+
+  // The watched descriptor returned from inotify_add_watch().
+  int mWatchedResourceDescriptor;
+
+  WatchedResourceDescriptor(const nsAString& aPath, const int anHandle)
+    : mPath(aPath)
+    , mWatchedResourceDescriptor(anHandle)
+  {
+  }
+};
+
+/**
+ *  * A structure used to pass the callbacks to the AddPathRunnableMethod() and
+ *   * RemovePathRunnableMethod().
+ *    */
+struct PathRunnablesParametersWrapper
+{
+  nsString mPath;
+  nsMainThreadPtrHandle<nsINativeFileWatcherCallback> mChangeCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback> mErrorCallbackHandle;
+  nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>
+    mSuccessCallbackHandle;
+
+  PathRunnablesParametersWrapper(
+    const nsAString& aPath,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherCallback>& aOnChange,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherErrorCallback>& aOnError,
+    const nsMainThreadPtrHandle<nsINativeFileWatcherSuccessCallback>&
+      aOnSuccess)
+    : mPath(aPath)
+    , mChangeCallbackHandle(aOnChange)
+    , mErrorCallbackHandle(aOnError)
+    , mSuccessCallbackHandle(aOnSuccess)
+  {
+  }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by continuously calling the blocking
+ * function GetQueuedCompletionStatus. This function queries the status of the
+ * Completion I/O port initialized in the main thread. The watched resources are
+ * registered to the completion I/O port when calling |addPath|.
+ *
+ * Instead of using a loop within the Run() method, the Runnable reschedules
+ * itself by issuing a NS_DispatchToCurrentThread(this) before exiting. This is
+ * done to allow the execution of other runnables enqueued within the thread
+ * task queue.
+ */
+class NativeFileWatcherIOTask : public Runnable
+{
+public:
+  explicit NativeFileWatcherIOTask()
+    : Runnable("NativeFileWatcherIOTask")
+    , mShuttingDown(false)
+  {
+  }
+
+  NS_IMETHOD Run() override;
+  nsresult AddPathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult RemovePathRunnableMethod(
+    PathRunnablesParametersWrapper* aWrappedParameters);
+  nsresult DeactivateRunnableMethod();
+
+private:
+  // Maintain 2 indexes - one by resource path, one by resource |HANDLE|.
```

Comment:
Here, too, there's no such thing as `HANDLE` under Linux. Could you update the doc?